File: src/08.dbpointer/29.141.ts
import { ListNode } from "./11.83";

export function hasCycle(head: ListNode | null): boolean {
    //     -----
    //    |     |
    // 3, 2, 0, 4
    // f
    // l
    // ------------
    //       f
    //    l  --------
    // ------------  |
    //    f  --------    
    //       l 

    let slow = head
    let fast = head

    while (fast) {
        fast = fast.next
        if (fast) {
            fast = fast.next
        }

        if (fast == slow) {
            return true
        }
        slow = slow.next
    }

    return false;
}

File: src/08.dbpointer/29.141.go
package dbpointer

func HasCycle(head *ListNode) bool {
	slow := head
	fast := head

	for fast != nil {
		fast = fast.Next
		if fast != nil {
			fast = fast.Next
		}

		if fast == slow {
			return true
		}

		slow = slow.Next
	}

	return false
}
